server/services/survey.service.js

```diff
@@ -0,0 +1,60 @@
+const Path = require('path-parser');
+const { URL } = require('url');
+
+module.exports = class SurveyService {
+
+  async updateSurveys(data) {
+    try {
+
+      const uniqueServeys =
+          data.map(this.buildSurveyDataFromBody)
+              .filter(this.filterOutNullAndUndefined)
+              .filter(this.filterByUniqueEmailAndSurveyId);
+
+      for (let survey of uniqueServeys) {
+        await Survey.updateOne(
+            {
+              _id: survey.surveyId,
+              recipients: {
+                $elemMatch: {email: survey.email, responded: false}
+              }
+            },
+            {
+              $inc: {[survey.choice]: 1},
+              $set: {'recipients.$.responded': true},
+              lastResponded: new Date()
+            }
+        ).exec();
+      }
+
+      return { success: true }
+    } catch(error) {
+      return error;
+    }
+  }
+
+  getRecipientListFromEmailString(recipients) {
+    return recipients.split(',').map(email => {
+      return { email };
+    })
+  }
+
+
+  buildSurveyDataFromBody(bodyItem) {
+    const path = new Path('/api/surveys/:surveyId/:choice');
+    const match = path.test(new URL(surveyData.url).pathname);
+    if (match) {
+      return { email: surveyData.email, surveyId: match.surveyId, choice: match.choice };
+    }
+  }
```

Comment:
Just noticed an error from when I moved code around. I call the passed in object `bodyItem`, but then later call it `surveyData`. I forgot to fix this after moving it into it's own service. They should be the same thing.

This is called by the map function (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map). Map just runs on an array and returns a new array based on the function passed in. So here we had an array of objects with emails and urls, and we turn it into an array of objects with emails, surveyIds, and choices.